#257 Support JSON References in contexts not allowed by Swagger 2.0

tfesenko opened the pull request:
#257 User-configurable severity for unsupported $ref context 
See proposed GUI description inZEN-3180. 
New preferences page:
<img width="660" alt="screen shot 2017-01-18 at 12 44 34 pm" src="https://cloud.githubusercontent.com/assets/644582/22081733/bfa7aa7c-dd92-11e6-8cb2-0271f2adc7ce.png">

Screencast:
![validation_ref](https://cloud.githubusercontent.com/assets/644582/22084826/8dd11742-dd9f-11e6-9b78-cae48df48fd8.gif)


Swagger spec used in the screencast demo:
```yaml
swagger: '2.0'
info:
  version: '2.0.0'
  title:  Some API
  description: Provides an interface for something
basePath: /usage/2
consumes:
  - application/json
produces:
  - application/json
    
paths: 
  /resource:
    get:
 # Security Requirement Object
      security: 
        - $ref: './security.yaml#/security/0' 
      responses:
        '200':
          description: OK

# Security Requirement Object
security: 
  - api_key_query_token: []
  - api_key_header: []
  - $ref: './security.yaml#/security/0' 

# Security Requirements Array
#security: 
#  $ref: './security.yaml#/security'  
 

#  Security Scheme Object
#securityDefinitions: 
#  api_key_query_token:
#    $ref: './security.yaml#/securityDefinitions/api_key'   

# Security Definitions Object
securityDefinitions: 
  $ref: './security.yaml#/securityDefinitions'   
```

Referenced security.yaml:
```yaml
swagger: '2.0'
info:
  version: '2'
  title: Common Security Objects
paths: {}

securityDefinitions:

  api_key:
    type: apiKey
    name: api_key
    in: header

  petstore_auth:
    type: oauth2
    authorizationUrl: http://swagger.io/api/oauth/dialog
    flow: implicit
    scopes:
      write:pets: modify pets in your account
      read:pets: read your pets
  
security:
  - api_key: []
  - petstore_auth: []

```

tfesenko committed:
#257, ZEN-3180 User-configurable severity for unsupported $ref context 

Dynamically modify the JSON Schema to allow $ref as a oneOf alternative
to the normal contents of one of the supported objects.

GUI is a group of checkboxes with the following labels:
Allow JSON references in additional contexts:
Security Definitions Object
Security Scheme Object
Security Requirements Array
Security Requirement Object

The JSON Schema was modified to allow dynamically added elementRefs

tfesenko committed:
#257 - group the validation preference controls together

tfesenko committed:
#257 - set proper default values for the validation preferences + fix ClassCast

tfesenko committed:
#257 - change look-n-feel of the Validation Preferences page

tfesenko committed:
#257 - fix augmentation of the JSON Schema for Swagger

SwaggerEditor listens to the notifications about preference changes:
1. Update the JSON Schema for Swagger according to the new settings.
2. Validate the open Swagger spec based on the new settings.

If several Swagger specs are open each editor will receive a
notification and try to update the JSON Schema. This commit prevents
multiple updates of the same kind, e.g. adding a refToJsonReference to
the same definition several times.

tfesenko committed:
#257 - fix validation preference for Security Definitions Object

Security Definitions Object is a special cases because it uses
`additional` properties to define values of the keys. I

tfesenko committed:
#257 - revert unneeded changes to Swagger Schema introducde in
56e61a120195ba4ae827119f142fde419f9cb91c

tfesenko committed:
#257 - fix labels for the new  appearance that was previously causing a failure in MultipleSwaggerErrorMessageTest

tedepstein commented on com.reprezen.swagedit/src/com/reprezen/swagedit/schema/schema.json:
```diff
@@ -1158,18 +1158,24 @@
+	       }]
+	     }  
+	   }
+	   ],
```
Clever!

tedepstein reviewed (commented):
Looked through the code.  It took me a few minutes to digest the approach you took here, but it makes sense.  Nice!

andylowry commented on com.reprezen.swagedit/src/com/reprezen/swagedit/schema/SwaggerSchema.java:
```diff
@@ -143,6 +147,40 @@ protected void init() {
+            }
+        }
     }
 
```
Very cool solution!

andylowry commented on com.reprezen.swagedit/src/com/reprezen/swagedit/preferences/SwaggerValidationPreferences.java:
```diff
@@ -31,6 +33,7 @@
         implements IWorkbenchPreferencePage, IPropertyChangeListener {
 
     public SwaggerValidationPreferences() {
+        // GRID is needed because we are not attaching the editor fields directly to FieldEditorParent, but to its child
```
This does not make the group work properly in FieldEditorPreferencePage. The problems are much deeper than that. It works properly in certain conditions, but not all. Adding another group or fields with different column requirements will likely lead to unfixable format issues, which is why I'd recommend a comment to this effect here. This breaks the assumption of the FEPP class.

The eclipse documentation on FEPP is misleading on this. It does not point out the risks of using your own layouts, and in fact the FEPP javadoc invites you to override #createLayout if you have special layout needs. This article is closer to correct: http://www.eclipse.org/articles/Article-Field-Editors/field_editors.html. Still, it makes it sound like you can use your own field editors at liberty within your own custom layout. Even that's not strictly true. The bottom line is that FieldEditors all manipulate the containing grid's layout, in ways that cannot be managed properly. So you can use FieldEditors within your own layout, as long as that layout puts every group of "sibling" field editors under the control of the same GridLayout, with nothing else in that GridLayout.

tedepstein replied:
@tfesenko , @andylowry , BTW, if we were to start this over again, I might have suggested using a checked listbox instead of a group of static listbox controls.  Only because it makes it easier (I think) to add new contexts. 

If you think the edge cases, where the current layout might break, warrant a fix, consider using a checked list as a possible alternative.  But at this point, please choose whichever fix is easiest.  (Or choose no fix at all, if you think we're OK with the current implementation.)

andylowry replied:
In that case it couldn't be a FieldEditorPreference anyway, unless we implemented a ListBoxFieldEditor (or whetever), since there's no existing FieldEditor for that sort of control.

andylowry commented on com.reprezen.swagedit/src/com/reprezen/swagedit/schema/SwaggerSchema.java:
```diff
@@ -171,13 +173,15 @@ public void allowJsonRefInContext(String jsonReferenceContext, boolean allow) {
             throw new IllegalArgumentException("Invalid JSON Reference Context: " + jsonReferenceContext);
         }
-        JsonNode refNode = mapper.createObjectNode().put("$ref", "#/definitions/elementRef");
+        int lastIndex = definition.size() - 1;
```
I'm not crazy about the assumption that this will always be the last oneOf element. I know that in current code it will be, but perhaps it's an assumption that will be unwittingly violated in some future code, leading to sporadic failures. Might be better to do a List#contains test on the entire oneOf array. Slightly slower, but this only happens when preferences change.

tfesenko replied:
@andylowry , I addressed your suggestion in  74446fa

andylowry commented on com.reprezen.swagedit/src/com/reprezen/swagedit/editor/SwaggerEditor.java:
```diff
@@ -191,7 +191,8 @@ public void propertyChange(PropertyChangeEvent event) {
                 }
             }
```
Test - please ignore

tfesenko commented on com.reprezen.swagedit/src/com/reprezen/swagedit/preferences/SwaggerValidationPreferences.java:
```diff
@@ -42,17 +45,23 @@ public void init(IWorkbench workbench) {
-        group.setText("Allow JSON references in additional contexts:");
+        // IMPORTANT: FieldEditorPreferencePage does not work very well with complex layouts and nested widgets.
+        // Consider switching to com.modelsolv.reprezen.generators.ui.preferences.GroupedFieldEditorPreferencePage from
+        // the RepreZen repo before modifying it
```
@andylowry , the comment with suggestions for future modifications is here

tfesenko committed:
#257 - prevent potential 

The modified code had the assumption that this would always be the last
oneOf element. This assumption can be unwittingly violated in some
future code, leading to sporadic failures. This change protects us from
this kind of problems in future. The current code is slightly slower,
but this only happens when preferences change.

andylowry reviewed (commented):
@tfesenko Looks good.
